Tox, for parallel test goodness

efischer19 opened the pull request:
# [EDU-768](https://openedx.atlassian.net/browse/EDUCATOR-768)

Attempts to split each of the following into separate test contexts on travis:
- django 1.8 python tests
- django 1.11 python tests
- quality checks
- js tests

efischer19 commented:
This has now morphed into a "modernize ORA" PR. With the goal of faster, smaller, parallelized tests, I'll be doing the following:

- [ ] modify installation, there's no reason we need to specify apt-packages. Requirements were also getting silly.
- [ ] remove AI grading, that eats a lot of testing cycles for a feature that isn't enabled anywhere. I'll double-check with product to ensure this is acceptable.
- [ ] tests will be more direct, and not make use of "travis calls tox calls makefile calls a bash script"

efischer19 commented:
Running into an order-dependent flaky test in test_training.py, under `--hashseed 2507972091`

efischer19 commented on requirements/tox.txt:
```diff
@@ -0,0 +1,24 @@
+django-nose==1.4.4
+git+https://github.com/edx/django-rest-framework.git@1ceda7c086fddffd1c440cc86856441bbf0bd9cb#egg=djangorestframework==3.6.3
+dogapi==1.11.1
+jsonfield==2.0.2
```
@MichaelRoytman just saw your recent branch - I'm upgrading `jsonfield` here as part of https://openedx.atlassian.net/browse/EDUCATOR-768

efischer19 commented on requirements/tox.txt:
```diff
@@ -0,0 +1,22 @@
+# edX Internal Requirements
+git+https://github.com/edx/XBlock.git@efischer/play_nice_with_py27#egg=XBlock==1.0.0
```
Relies on https://github.com/edx/XBlock/pull/365, which is currently unmerged.

efischer19 replied:
Merged and updated

efischer19 commented:
Ready for review! @cahrens has volunteered to be a reviewer (thanks again!); @edx/educator-dahlia FYI as well.

Other subtasks before I close EDUCATOR-409:

- [ ] upload to PyPI, I'll do this work in a separate PR for easier reviewing
- [ ] get https://github.com/edx/XBlock/pull/365 merged
- [ ] update Confluence docs for install/test/developing against edx-ora2

cahrens commented on logs/README.txt:
```diff
@@ -1,7 +0,0 @@
-Log files:
-
-apps_info.log = INFO level logging for all edx-ora2 apps and OpenAssessmentBlock
```
Should this information be captured someplace else? It seems like it could be useful (assuming it is still true).

efischer19 replied:
I don't believe this info is true any longer, edx-platform defines logging locations (which work with splunk), and this particular file is >3 years past its last update.

cahrens commented on openassessment/assessment/models/ai.py:
```diff
@@ -1,874 +1 @@
-        )
-        AssessmentPart.create_from_option_points(self.assessment, criterion_scores)
-        self.mark_complete_and_save()
+# This file is empty, but we cannot delete it b/c historical migration records refer to it.
```
Is this still true given our discussion?

efischer19 replied:
I will double check on this one - it could be in a weird intermediary state from my internal debate about dealing with migrations.

cahrens commented on openassessment/assessment/tasks.py:
```diff
@@ -1,7 +0,0 @@
-"""
-# pylint:disable=W0611
-from .worker.training import train_classifiers, reschedule_training_tasks
-from .worker.grading import grade_essay, reschedule_grading_tasks
```
Why is worker.grading removed?

efischer19 replied:
Everything under worker has been removed - the idea was to "train" your AI classifiers with a given data set ahead of time, then "grade" new essays as a separate phase. We don't need any references to either of these operations.

cahrens commented on openassessment/fileupload/backends/swift.py:
```diff
@@ -32,10 +35,11 @@ def get_upload_url(self, key, content_type):
         try:
             temp_url = swiftclient.utils.generate_temp_url(
-                path='%s/%s/%s' % (url.path, bucket_name, key_name),
+                path='/v%s%s/%s/%s' % (SWIFT_BACKEND_VERSION, url.path, bucket_name, key_name),
```
Will this cause any incompatibilities for people who are already using the Swift backend?

efischer19 replied:
It shouldn't, this only affects newly generated upload urls.

cahrens commented on openassessment/management/commands/performance_test_for_submissions.py:
```diff
@@ -1,74 +0,0 @@
-from openassessment.assessment.models import PeerWorkflow
-
-
-class Command(BaseCommand):
```
Why was this file deleted?

efischer19 replied:
I yanked out all performance tests as well, as they also heavily blurred the line between "python library" and "suite of scripts and programs to install, some of which may be python". The tests that were here were incredibly old, I think if we need performance testing against ora we're better off rewriting it in https://github.com/edx/edx-load-tests using the modern locust setup described there.

cahrens commented on openassessment/management/tests/test_upload_oa_data.py:
```diff
@@ -69,4 +67,4 @@ def test_upload(self):
         # Expect that we generated a URL for the bucket
         url = cmd.history[0]['url']
-        self.assertIn("https://{}".format(self.BUCKET_NAME), url)
+        self.assertIn("ttps://s3.amazonaws.com/{}".format(self.BUCKET_NAME), url)
```
Why is the "h" missing from the front?

efischer19 replied:
Terrible copy-paste job by the dev who put this here, what a bum that guy must be (thanks, will fix)

cahrens commented on openassessment/tests/test_data.py:
```diff
@@ -363,13 +363,15 @@ class TestOraAggregateDataIntegration(TransactionCacheResetTest):
         self.possible_points = 2
         peer_api.get_submission_to_assess(self.scorer_submission['uuid'], 1)
         self.assessment = self._create_assessment(self.scorer_submission['uuid'])
+        self.assertEqual(self.assessment['parts'][0]['criterion']['label'], "criterion_1")
```
What caused this change (and the lines below, 475 and following?

efischer19 replied:
It was a flaky test issue. A particular hashseed value caused the assessment to store out-of-order values, until I made [this fix](https://github.com/edx/edx-ora2/pull/1023/files/5d5921287a9c00b3935b86756cd960f64374b678#diff-c09750592365024b37e7601ab2b68bad). Now, this test ensures that `self.assessment['parts'][0]` matches up to the first criterion and appears first in the downloaded data.

cahrens commented on openassessment/xblock/static/sass/oa/views/_oa-base.scss:
```diff
@@ -1086,7 +1086,7 @@
 
     .action--save {
-      @extend %action-button
+      @extend %action-button !optional
```
Why was this necessary?

efischer19 replied:
It was failing quality builds, not sure why it hadn't been before. I'll double-check to ensure that's still the case.

efischer19 replied:
Example failure: https://travis-ci.org/edx/edx-ora2/jobs/257446578

cahrens commented on openassessment/xblock/static/sass/oa/views/_oa-base.scss:
```diff
@@ -1086,7 +1086,7 @@
+      @extend %action-button !optional
     }
 
     .feedback {
```
Is there CSS that should be deleted related to AI grading? I'm thinking of the Studio editors, as well as the LMS view.

efischer19 replied:
Oh, good call. I'll do an audit of that before the next round of review

efischer19 replied:
I've poked around - no files are obviously AI grading exclusive, and I can't see any obvious removals in the existing files. Judging by that and the fact that very little HTML needed to be removed, I think we're safe assuming that this feature never had a lot of dedicated front-end customization done.

cahrens commented on openassessment/xblock/studio_mixin.py:
```diff
@@ -379,19 +347,13 @@ def _editor_assessments_order_context(self):
-        # If the problem already contains example-based assessment
-        # then allow the editor to display example-based assessments,
-        # which is not included in the default
-        enabled_assessments = [asmnt['name'] for asmnt in self.valid_assessments]
```
Have you checked that no ORA instances in prod/Edge have AI steps defined? Just wondering how they would fail.

efischer19 replied:
From my database poking, I don't believe there are any. I'll verify that, and set up such a problem locally to ensure we can handle that situation correctly.

cahrens commented on scripts/run-pep8.sh:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-MAX_PEP8_VIOLATIONS=106
+MAX_PEP8_VIOLATIONS=62  # TODO nomerge - but also stop failing during development
```
Don't forget to follow up on this.

cahrens commented on scripts/run-pylint.sh:
```diff
@@ -1,6 +1,6 @@
 #!/usr/bin/env bash
 
-MAX_PYLINT_VIOLATIONS=438
+MAX_PYLINT_VIOLATIONS=495  # TODO nomerge - but also stop failing during development
```
Don't forget to follow up on this.

cahrens commented on test/performance/README.rst:
```diff
@@ -1,50 +0,0 @@
-Performance Tests
-=================
-
-1. Install performance test requirements:
```
No more performance tests?

efischer19 replied:
Correct, I've eliminated them from this repo for reasons outlined above.

cahrens commented:
What are the .gitkeep files?

cahrens commented:
What is the situation with the course.tar.gz file (I don't understand the status from the files tab)?

cahrens commented:
@efischer19 I'm done slogging through this-- wow, a lot of great cleanup!

efischer19 commented:
`.gitkeep` files [are placed in a directory to ensure it exists on checkout](https://stackoverflow.com/questions/7229885/what-are-the-differences-between-gitignore-and-gitkeep). I thought those directories were overkill and have removed them, so the files are deleted.

course.tar.gz is a binary file that had been included for performance testing. I think storing it in git is the wrong move, but I also just realized that removing it may cause problems with https://tools-edx-jenkins.edx.org/job/ora2/job/add-course-to-sandbox/. I'll add that to my list of things to look into.

Thanks for the review! I'll let you know when I've addressed everything.

cahrens commented:
@efischer19 related to the .gitkeep files, I was confused by this "grey" state. What does it mean?

![image](https://user-images.githubusercontent.com/484484/28544776-72c155ee-7092-11e7-95b9-45854266605e.png)

efischer19 commented:
Oh! I think that's github's UI having a hard time with "This file of size 0 has been deleted" - it does correctly show that no lines have changed, but https://github.com/edx/edx-ora2/blob/5d5921287a9c00b3935b86756cd960f64374b678/storage/dev/.gitkeep 404s, so there's no file there.

efischer19 commented:
The course that gets added to sandboxes for acceptance testing lives under `scripts/data`, the one I've deleted was under `test/performance`, so [jenkins can still do its job](https://tools-edx-jenkins.edx.org/job/ora2/job/add-course-to-sandbox/132)

efischer19 commented:
I've confirmed (locally) that a problem with a defined example-based assessment step will still render properly.

efischer19 commented:
Alright @cahrens, I believe this is ready again, I've addressed all your comments from before. The diff of the latest commit looks scary, but a large portion of it was me running `isort` to standardize all our import statements (which fixed a huge number of quality errors).

I'm also running acceptance tests at https://gocd.tools.edx.org/go/pipelines/value_stream_map/build_ora2_sandbox/142.

efischer19 commented on requirements/base.txt:
```diff
@@ -1,31 +1,23 @@
 # edX Internal Requirements
-git+https://github.com/edx/XBlock.git@xblock-0.4.12#egg=XBlock==0.4.12
+# edx-submissions>=2.0.6,<3.0.0 #TODO: update after 2.0.6 release
+git+https://github.com/edx/edx-submissions.git@efischer/loosen_reqs#egg=edx-submissions==2.0.6 # TODO: and remove this line
```
These two lines are effectively "I'm waiting for https://github.com/edx/edx-submissions/pull/65 to merge before this"

cahrens commented on Makefile:
```diff
@@ -2,6 +2,7 @@
 # Install commands
 ##################
 install-python:
+	pip install -r requirements/django.txt
```
Why a separate file?

cahrens replied:
I see-- because of the tox configuration. Never mind.

cahrens commented on openassessment/workflow/migrations/0002_auto_20170724_1537.py:
```diff
@@ -0,0 +1,21 @@
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
```
What caused this migration?

efischer19 replied:
It's the same thing that's happening [here](https://github.com/edx/edx-submissions/pull/65/files#r130444360) - We can remove the `django_extensions` dependency fairly easily, but that touches a field on a model, which leads to a migration. Said migration *should* be a no-op, but there are some pending operational questions to sort out.

The migration should have been in the last commit; I had forgotten that it needed to be generated and ran into a "please run makemigrations" warning while fixing other things. I'll work out the details with Kevin on my other PR and update this one with the results of that conversation.

efischer19 replied:
Update: we're pushing the submissions upgrade through independently, as it had an identical migration. Expectation is that it runs as a no-op, we're going to verify that as submissions goes out since it's a smaller table.

cahrens commented on requirements/base.txt:
```diff
@@ -1,3 +1,23 @@
+# edX Internal Requirements
+# edx-submissions>=2.0.6,<3.0.0 #TODO: update after 2.0.6 release
+git+https://github.com/edx/edx-submissions.git@efischer/loosen_reqs#egg=edx-submissions==2.0.6 # TODO: and remove this line
+git+https://github.com/edx/django-rest-framework.git@1ceda7c086fddffd1c440cc86856441bbf0bd9cb#egg=djangorestframework==3.6.3
```
Why a specific commit of django-rest-framework?

efischer19 replied:
@doctoryes set this requirement to a specific commit in https://github.com/edx/edx-ora2/pull/1024, I'm just moving it to another file

cahrens commented on setup.cfg:
```diff
@@ -0,0 +1,9 @@
+known_third_party = ddt,mock
+known_django = django
+known_djangoapp = jsonfield,model_utils
+known_edx =
```
known_edx?

efischer19 replied:
Yeah, it has something to do with isort configuration. I copied this over from https://github.com/edx/edx-celeryutils/blob/master/setup.cfg, I think we can remove the `known_edx` line

cahrens commented on settings/base.py:
```diff
@@ -93,6 +92,12 @@
     'django.template.loaders.filesystem.Loader',
```
These are the default values, so TEMPLATE_LOADERS can be removed.

https://docs.djangoproject.com/en/1.11/ref/templates/upgrading/

cahrens commented on settings/base.py:
```diff
@@ -3,7 +3,6 @@
-from celery import Celery
 
 DEBUG = True
 TEMPLATE_DEBUG = DEBUG
```
I believe you should remove TEMPLATE_DEBUG. You can include 'debug' in the TEMPLATES options.

For example,

````
TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [
            os_path.join(PROJECT_PATH, 'templates'),
        ],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                "django.contrib.auth.context_processors.auth",
                "django.template.context_processors.debug",
                "django.template.context_processors.i18n",
                "django.template.context_processors.media",
                "django.template.context_processors.static",
                "django.template.context_processors.request",
                "django.template.context_processors.tz",
                "django.contrib.messages.context_processors.messages",
                "sekizai.context_processors.sekizai",
            ],
            'debug': DEBUG,
        },
    },
]

cahrens replied:
Also, STATICFILES_DIRS and STATICFILES_FINDERS can be removed.

https://docs.djangoproject.com/en/1.11/ref/settings/#std:setting-STATICFILES_FINDERS

cahrens commented:
This seems very odd:

![image](https://user-images.githubusercontent.com/484484/28789854-0e3258e0-75f5-11e7-9d9b-bed4c5874212.png)

cahrens commented:
@efischer19 done re-reviewing-- just a few more comments/questions.

efischer19 commented:
Thanks for the review @cahrens! I think I've answered all your questions, will get this updated witch code changes and ready for merge in the next day or so.

efischer19 commented:
We're back - latest commit has:

- updated location of developer wiki links (I'm going to be editing those soon to explain docker devstack setup FYI)
- finalized edx-submissions requirement to 2.0.6, since that merged
- lowered pylint error limit, since I was well below it
- cleaned up settings/base.py as suggested
- preemptively fixed a flaky test I noticed earlier this morning

Once tests have passed and everyone is happy, I'll get this squashed. merged, and deployed

cahrens commented:
👍

efischer19 commented:
🎆 

Thanks for the help here @cahrens!

I'm going to get this deployed tomorrow

cahrens commented:
🚀
